m2cgen/interpreters/dart/code_generator.py

```diff
@@ -0,0 +1,54 @@
+import contextlib
+
+from m2cgen.interpreters.code_generator import CLikeCodeGenerator
+
+
+class DartCodeGenerator(CLikeCodeGenerator):
+
+    scalar_type = "double"
+    vector_type = "List<double>"
+
+    def __init__(self, *args, **kwargs):
+        super(DartCodeGenerator, self).__init__(*args, **kwargs)
+
+    def add_class_def(self, class_name, is_private=False):
```

Comment:
I have never heard about the Dart language before 🙂 , but according to the following articles
- https://en.wikipedia.org/wiki/Dart_(programming_language)
- https://dart.dev/guides/language/language-tour

it seems to me that classes are optional stuff in this language. I think we should try to generate code as simple as possible and here it can be simplified to an ordinary function (unlike in Java and C# where classes are essential part of the language). Please correct me if I'm wrong.

Reply:
I have only a little experience in Dart. I can definitely make this change. My only reservation with doing so is that class encapsulation makes sense as the linear algebra and tanh functions are added as helpers within the class and really do not need to be accessible outside (in which I could change the naming convention by prepending a '_' to keep them private). Also, it feels more natural accessing the RegressorModel class's functionality than just grabbing a function out of a file. If you could go either way, I would say the class method has a slight edge. At the end of the day, the developer using this will do whatever they please to the code to make it work in their project.

For what it is worth, Wikipedia defines the language as:
"Dart is an object-oriented, class defined, garbage-collected language..."
which sounds a lot like Java's definition:
"Java is a general-purpose programming language that is class-based, object-oriented..."

You let me know what the final say is, boss!

Reply:
Nowadays a lot of programming languages are multi-paradigm and we can say the same for many of them. For instance, we can use classes in JS, Python, PHP, PowerShell, etc. but we don't.

Anyway, @izeigerman will have the last word.

Reply:
I want to make sure I conform my code to what the contributors of this repo want, so I will do whatever you want.

I did come across this article for writing effective Dart code: https://dart.dev/guides/language/effective-dart/design#avoid-defining-a-class-that-contains-only-static-members

Looks like you have the upper hand 😉.

I will make the necessary changes as soon as I can.

Reply:
My next question would be, does using the Dart `library` construct mentioned in that article make sense? Could be a good way to scope the model or provide some semantic sugar upon importing the code.

I do not really know how the typical user uses the tool. I would guess devs will export the code and then rename things and paste the code where they want it as they see fit, so the extra code just causes headaches on our end to maintain it.

Reply:
Thank you, folks. 
I'm a bit biased in this regard since I'm not a huge fan of the object-oriented paradigm and find classes to be redundant. Though in some languages they are the only way to achieve modularity, and hence a necessary evil.  

Long story short, since the code generated by `m2cgen` doesn't maintain any state that requires encapsulation plus Dart has first-class functions - I vote to drop classes. 

> My next question would be, does using the Dart library construct mentioned in that article make sense? Could be a good way to scope the model or provide some semantic sugar upon importing the code.

Quoting the source that you shared:
> If what you want is a namespace, a library is a better fit. Libraries support import prefixes and show/hide combinators. Those are powerful tools that let the consumer of your code handle name collisions in the way that works best for them.

So it seems like Dart leaves a lot to the consumer's consideration so why shouldn't we. Let's leave it up to a user to decide how to handle the generated code. Additionally having a function name parameterized should help with name collision. This feature is currently being introduced in the adjacent PR (https://github.com/BayesWitnesses/m2cgen/pull/166). Perhaps we should consider proactively introducing this parameter to the Dart interpreter.